test_conformance/extensions/cl_khr_kernel_clock/kernel_clock.cpp

```diff
@@ -0,0 +1,198 @@
+                    sprintf(kernel_src, kernel_sources[i], "device", "device");
+                    break;
+                }
+                case CL_DEVICE_KERNEL_CLOCK_SCOPE_WORK_GROUP_KHR: {
+                    sprintf(kernel_src, kernel_sources[i], "workgroup",
+                            "workgroup");
+                    break;
+                }
+                case CL_DEVICE_KERNEL_CLOCK_SCOPE_SUB_GROUP_KHR: {
+                    sprintf(kernel_src, kernel_sources[i], "subgroup",
+                            "subgroup");
+                    break;
+                }
+            }
+
+            ptr = kernel_src;
+
+            error = create_single_kernel_helper_create_program(
+                context, &program, 1, &ptr);
+            test_error(error, "Failed to create program with source");
+
+            error =
+                clBuildProgram(program, 1, &device, nullptr, nullptr, nullptr);
+            test_error(error, "Failed to build program");
+
+            out_mem = clCreateBuffer(context, CL_MEM_WRITE_ONLY,
+                                     sizeof(cl_uint), nullptr, &error);
+            test_error(error, "clCreateBuffer failed");
+
+            kernel = clCreateKernel(program, "SampleClock", &error);
+            test_error(error, "Failed to create kernel");
+
+            error = clSetKernelArg(kernel, 0, sizeof(out_mem), &out_mem);
+            test_error(error, "clSetKernelArg failed");
+
+            error = clEnqueueNDRangeKernel(queue, kernel, 1, NULL, &global_size,
+                                           NULL, 0, NULL, NULL);
+            test_error(error, "clNDRangeKernel failed");
+
+            error = clEnqueueReadBuffer(queue, out_mem, CL_BLOCKING, 0,
+                                        sizeof(cl_uint), &buf, 0, NULL, NULL);
+            test_error(error, "clEnqueueReadBuffer failed");
+
+            if (buf == 1)
+            {
+                log_error(
+                    "Sampling the clock returned bad values, time1 > time2.\n");
+                return TEST_FAIL;
+            }
+        }
+
+        return CL_SUCCESS;
+    }
+
+private:
+    cl_device_id device;
+    cl_context context;
+    cl_command_queue queue;
+    cl_device_kernel_clock_capabilities_khr capability;
+};
+
+int MakeAndRunTest(cl_device_id device, cl_context context,
+                   cl_command_queue queue,
+                   cl_device_kernel_clock_capabilities_khr capability)
+{
+    if (!is_extension_available(device, "cl_khr_kernel_clock"))
+    {
+        log_info(
+            "The device does not support the cl_khr_kernel_clock extension.\n");
+        return TEST_SKIPPED_ITSELF;
+    }
+
+    KernelClockTest test_fixture =
+        KernelClockTest(device, context, queue, capability);
+
+    if (test_fixture.Skip())
+    {
+        return TEST_SKIPPED_ITSELF;
+    }
+
+    cl_int error = test_fixture.RunTest();
+    test_error_ret(error, "Test Failed", TEST_FAIL);
+
+    return TEST_PASS;
+}
```

Comment:
You can probably wrap everything until here in an anonymous namespace.

Reply:
Done